packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -3,32 +3,50 @@
  * Licensed under the MIT License.
  */
 
-import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { ITelemetryLogger, ITelemetryProperties } from "@fluidframework/common-definitions";
 import { IConnectionDetails } from "@fluidframework/container-definitions";
 import { ConnectionMode, IQuorumClients, ISequencedClient } from "@fluidframework/protocol-definitions";
 import { PerformanceEvent } from "@fluidframework/telemetry-utils";
 import { assert, Timer } from "@fluidframework/common-utils";
 import { ConnectionState } from "./connectionState";
 
-export interface IConnectionStateHandler {
+export interface IConnectionStateHandlerProps {
+    /** Provides access to the clients currently in the quorum */
     quorumClients: () => IQuorumClients | undefined;
-    logConnectionStateChangeTelemetry: (
-        value: ConnectionState,
-        oldState: ConnectionState,
-        reason?: string | undefined
-    ) => void;
+    /** Log to telemetry any change in state, included to Connecting */
+    logConnectionStateChangeTelemetry:
+        (value: ConnectionState, oldState: ConnectionState, reason?: string | undefined) => void;
+    /** Whether to expect the client to join in write mode on next connection */
     shouldClientJoinWrite: () => boolean;
+    /** (Optional) How long should we wait on our previous client's Leave op before transitioning to Connected again */
     maxClientLeaveWaitTime: number | undefined;
-    logConnectionIssue: (eventName: string) => void;
+    /** Log an issue encountered while in the Connecting state. details will be logged as a JSON string */
+    logConnectionIssue: (eventName: string, details?: ITelemetryProperties) => void;
+    /** Callback whenever the ConnectionState changes between Disconnected and Connected */
     connectionStateChanged: () => void;
 }
 
 export interface ILocalSequencedClient extends ISequencedClient {
     shouldHaveLeft?: boolean;
 }
 
-const JoinOpTimer = 45000;
+const JoinOpTimeoutMs = 45000;
 
+/**
+ * In the lifetime of a container, the connection will likely disconnect and reconnect periodically.
+ * Due to the distributed nature of the ordering service, the transition from old clientId to new clientId
+ * (since each reconnect gets a unique clientId) is asynchronous and the sequence of events is unpredictable.
```

Comment:
This class ensures that any ops sent by this container instances on previous connection are sequenced or will never get sequenced before allowing runtime to resubmit ops (i.e. know that connection was established)
clientId switching (proper timing) is result of the above - we need to ensure we keep old clientId in place while there is any chance for ops to come in (be acked) that were sent by this client using old connection - keeping old clientId ensures we will tread such ops as "local"